### PR TITLE
Hw5 Lecture 5 - BDD

### PR DESCRIPTION
**Pull request by solomencev**

### CODE REVIEW

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Given("User opens website")
+    public void openWebSite() {
+        mainPage.openMainPage();
+    }
+
+    @Given("I open JDI GitHub site")
+    public void openJdiGitHubWebsite() {
+        mainPage.openMainPage();
+    }
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: Please, choose "User" or "I" for definition of your steps :)

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    @When("Open through the header menu Service -> Different Elements Page")
+    public void openDifferentElementsPage() {
+        headerMenu.clickServiceMenuAndThenDifferentElementService();
+    }
+
+    @When("I click on {string} button in Header")
+    public void openServiceMenuInHeader(String string) {
+        headerMenu.clickServiceMenu();
+    }
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: What is "String string"? Do you need this parameter?

File: src/main/java/com/epam/tc/hw5/fragments/HeaderMenu.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    @FindBy(xpath = "//a[text()='User Table ']")
+    private WebElement userTablePageElement;
+
+    public HeaderMenu(WebDriver driver) {
+        super(driver);
+    }
+
+    public void clickServiceMenuAndThenDifferentElementService() {
+        serviceMenuDropDown.click();
+        differentElementsServiceElement.click();
```
Comment (Dr-Java) [inline src/main/java/com/epam/tc/hw5/fragments/HeaderMenu.java]: I don't love this very long method name... clickServiceMenuAndThenDifferentElementService()

What do you want to do? Open another page? Or smth else?

Please, choose better name)

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    @When("Select radio Selen")
+    public void selectRadioSelen() {
+        differentElementsPage.selectRadioButtonSelen();
+    }
+
+    @When("Select in dropdown Yellow")
+    public void selectDropDownColorYellow() {
+        differentElementsPage.selectDropDownYellow();
+    }
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: I select OR User selects

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+public class AssertStep extends DriverSetup {
+
+    SoftAssertions softAssert = new SoftAssertions();
+    MainPage mainPage = new MainPage(driver);
+    DifferentElementsPage differentElementsPage = new DifferentElementsPage(driver);
+    UserTablePage userTablePage = new UserTablePage(driver);
+
+    @Then("Browser title equals 'Home Page'")
+    public void assertBrowserTitleHomePage() {
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: Browser title should be...

Please, do it for all your steps

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    }
+
+    @Then("{string} page should be opened")
+    public void assertBrowserTitleUserTable(String title) {
+        softAssert.assertThat(mainPage.getBrowserTitleName()).as("Incorrect title").isEqualTo(USER_TABLE_TAB);
+        softAssert.assertAll();
+    }
+
+    @Then("User is logged as {string}")
+    public void user_is_logged_as(String userFullName) {
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: Why not camel case? "user_is_logged_as"

File: src/main/java/com/epam/tc/hw5/fragments/HeaderMenu.java (hunk 3 above)
Comment (solomencev) [reply src/main/java/com/epam/tc/hw5/fragments/HeaderMenu.java]: The method opens the 'Different Element' page. The method has been renamed to openDifferentElementPage()

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java (hunk 2 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: This parameter has been removed. The description of both feature files has been changed.

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java (hunk 6 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: The method has been renamed.

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java (hunk 5 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: Updated the name of all methods and their description

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java (hunk 4 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: Updated the name of all methods and their description

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java (hunk 1 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: Updated the name of all methods and their description

File: src/main/java/com/epam/tc/hw5/fragments/HeaderMenu.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    public void clickServiceMenu() {
+        serviceMenuDropDown.click();
+    }
+
+    public void clickUserTablePageElement() {
+        userTablePageElement.click();
+    }
+
+
```
Comment (Dr-Java) [inline src/main/java/com/epam/tc/hw5/fragments/HeaderMenu.java]: Please, remove extra lines.

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    HeaderMenu headerMenu  = new HeaderMenu(driver);
+    UserTablePage userTablePage = new UserTablePage(driver);
+
+    @Given("I open JDI GitHub site")
+    public void openJdiGithubSite() {
+        mainPage.openMainPage();
+    }
+
+    @When("Perform login as {string} and password {string}")
+    public void performLogin(String user, String password) {
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: "I perform login..."

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        mainPage.login(user, password);
+    }
+
+    @When("Open through the header menu Service -> Different Elements Page")
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: "I open..."

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+public class AssertStep extends DriverSetup {
+
+    SoftAssertions softAssert = new SoftAssertions();
+    MainPage mainPage = new MainPage(driver);
+    DifferentElementsPage differentElementsPage = new DifferentElementsPage(driver);
+    UserTablePage userTablePage = new UserTablePage(driver);
+
+    @Then("Browser title should be 'Home Page'")
+    public void assertUserTablePageIsOpened() {
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: What do you check? What page should be opened?

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        softAssert.assertThat(userTablePage.getBrowserTitleName()).as("Incorrect title").isEqualTo(HOME_PAGE);
+        softAssert.assertAll();
+    }
+
+    @Then("{string} page should be opened")
+    public void assertHomePageIsOpened(String title) {
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: What page should be opened? In your feature file you say "Then "User Table" page should be opened"... And why don't you use "String title" parameter?

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    private static boolean logsContainsItem(Collection<WebElement> logs, String expectedEntry) {
+        for (var logElement : logs) {
+            if (logElement.getText().contains(expectedEntry)) {
+                return true;
+            }
+        }
+        return false;
+    }
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: Is it possible to place boolean logsContainsItem method in Different Elements page class?

File: src/main/java/com/epam/tc/hw5/MainPage.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    @FindBy(id = "login-button")
+    private WebElement loginSubmit;
+
+    @FindBy(id = "user-name")
+    private WebElement loginName;
+
+    @FindBy(css = ".uui-navigation.nav.navbar-nav.m-l8>li")
+    private List<WebElement> itemsInHeader;
+
+    public static final String ID_FRAME = "frame";
```
Comment (Dr-Java) [inline src/main/java/com/epam/tc/hw5/MainPage.java]: Field 'ID_FRAME' is never used.

File: src/main/java/com/epam/tc/hw5/MainPage.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        loginSubmit.click();
+    }
+
+    public void switchToFrame(String frame) {
+        driver.switchTo().frame(frame);
+    }
+
+    public void switchToHomePage() {
+        driver.switchTo().defaultContent();
+    }
```
Comment (Dr-Java) [inline src/main/java/com/epam/tc/hw5/MainPage.java]: YAGNI =)

File: src/main/java/com/epam/tc/hw5/MainPage.java [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        return frame;
+    }
+
+    public List<WebElement> getLeftMenu() {
+        return leftMenu;
+    }
+
+    public WebElement getFrameButton() {
+        return frameButton;
+    }
```
Comment (Dr-Java) [inline src/main/java/com/epam/tc/hw5/MainPage.java]: YAGNI =)

File: src/main/java/com/epam/tc/hw5/UserTablePage.java [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    public void selectVipCheckbox(String userName) {
+        driver.findElement(By.xpath(".//*[contains(text(),'" + userName + "')]/../..//input"))
+              .click();
+    }
+
+    public boolean logsAfterSelectingVipCheckbox(String logText) {
+        By logPath = new By.ByXPath("//li[contains(. ,'" + logText + "')]");
+        WebElement log = driver.findElement(logPath);
+        return log.isDisplayed();
+    }
```
Comment (Dr-Java) [inline src/main/java/com/epam/tc/hw5/UserTablePage.java]: I do not love this naming for boolean method "boolean logsAfterSelectingVipCheckbox(String logText)"... We've discussed it in previous reviews))

File: src/main/java/com/epam/tc/hw5/fragments/HeaderMenu.java (hunk 7 above)
Comment (solomencev) [reply src/main/java/com/epam/tc/hw5/fragments/HeaderMenu.java]: removed extra lines

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java (hunk 8 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: Updated to 'I perform login'

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java (hunk 9 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: updated - _I open_ through the header menu Service -> Different Elements Page

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java (hunk 10 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: In this method, the opening of the Home page is checked.
Updated:
**_assertHomePageIsOpened_** () 
**_assertUserTableIsOpened_** ()

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java (hunk 11 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: Removed parameter in assertHomePageIsOpened()

File: src/main/java/com/epam/tc/hw5/MainPage.java (hunk 13 above)
Comment (solomencev) [reply src/main/java/com/epam/tc/hw5/MainPage.java]: Removed constant ID_FRAME

File: src/main/java/com/epam/tc/hw5/MainPage.java (hunk 14 above)
Comment (solomencev) [reply src/main/java/com/epam/tc/hw5/MainPage.java]: Removed unused methods switchToFrame(), switchToHomePage()

File: src/main/java/com/epam/tc/hw5/MainPage.java (hunk 15 above)
Comment (solomencev) [reply src/main/java/com/epam/tc/hw5/MainPage.java]: Removed unused methods getMenuList(), getListOfImages(),  getListOfTextUnderImages(),getFrame(), getLeftMenu(), getFrameButton();

File: src/main/java/com/epam/tc/hw5/UserTablePage.java (hunk 16 above)
Comment (solomencev) [reply src/main/java/com/epam/tc/hw5/UserTablePage.java]: Renamed method **areLogs**DisplayedAfterSelectingVipCheckbox()

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java (hunk 12 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: I placed and renamed boolean areLogsContainsItem() method in Different Elements page

File: src/main/java/com/epam/tc/hw5/UserTablePage.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class UserTablePage extends BasePage {
+    
+    public static final String USER_TABLE_TAB = "User Table";
+    @FindBy(xpath = "//tbody//select")
+    public List<WebElement> listOfDropdowns;
```
Comment (Dr-Java) [inline src/main/java/com/epam/tc/hw5/UserTablePage.java]: private

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        headerMenu.openDifferentElementPage();
+    }
+
+    @When("I click on Service button in Header")
+    public void openServiceMenuInHeader() {
+        headerMenu.clickServiceMenu();
+    }
+
+    @When("I click on {string} button in Service dropdown")
+    public void openUserTablePage(String string) {
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: Still I see this "String string" parameter here...

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+        softAssert.assertThat(userTablePage.getDroplistValues(userName))
+              .as("Incorrect values in dropdown")
+              .isEqualTo(values);
+        softAssert.assertAll();
+    }
+
+    @Then("1 log row has {string} text in log section")
+    public void logRowHasTextInLogSection(String log) {
+        softAssert.assertThat(userTablePage.areLogsDisplayedAfterSelectingVipCheckbox(log)).as("Fail").isTrue();
+    }
```
Comment (Dr-Java) [inline src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: What is "Fail"?

**Review by Dr-Java — APPROVED** [review]

File: src/main/java/com/epam/tc/hw5/UserTablePage.java (hunk 17 above)
Comment (solomencev) [reply src/main/java/com/epam/tc/hw5/UserTablePage.java]: Removed unused variable

File: src/test/java/com/epam/tc/hw5/steps/ActionStep.java (hunk 18 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/ActionStep.java]: Deleted parameter in the openUserTablePage()

File: src/test/java/com/epam/tc/hw5/steps/AssertStep.java (hunk 19 above)
Comment (solomencev) [reply src/test/java/com/epam/tc/hw5/steps/AssertStep.java]: updated description in the as() method

**Pull request merged**